core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -0,0 +1,951 @@
+     * that should be associated with some field of some entry.
+     * CAS from 'expectedIntValue' to 'newIntValue' for field at specified offset
+     * of given int-field in the entries array
+     */
+    private boolean casEntriesArrayInt(
+        int intFieldIdx, OFFSET offset, int expectedIntValue, int newIntValue) {
+        return unsafe.compareAndSwapInt(entries,
+                Unsafe.ARRAY_INT_BASE_OFFSET + (intFieldIdx + offset.value) * Unsafe.ARRAY_INT_INDEX_SCALE,
+                expectedIntValue, newIntValue);
+    }
+
+    /**
+     * casEntriesArrayLong performs CAS of given two integers field of the entries ints array,
+     * that should be associated with some 2 consecutive integer fields of some entry.
+     * CAS from 'expectedLongValue' to 'newLongValue' for field at specified offset
+     * of given int-field in the entries array
+     */
+    private boolean casEntriesArrayLong(
+        int intStartFieldIdx, OFFSET offset, long expectedLongValue, long newLongValue) {
+        return unsafe.compareAndSwapLong(entries,
+                Unsafe.ARRAY_INT_BASE_OFFSET + (intStartFieldIdx + offset.value) * Unsafe.ARRAY_INT_INDEX_SCALE,
+                expectedLongValue, newLongValue);
+    }
+
+    /********************************************************************************************/
+    /*--------------- Methods for setting and getting specific key and/or value ----------------*/
+    // key or value references are the triple <blockID, position, length> encapsulated in long
+    // whenever version is needed it is an additional integer
+
+    /**
+     * Atomically reads the value version from the entry (given by entry index "ei")
+     * synchronisation with reading the value reference is not ensured in this method
+     * */
+    private int getValueVersion(int ei) {
+        return getEntryArrayFieldInt(entryIdx2intIdx(ei), OFFSET.VALUE_VERSION);
+    }
+
+    /**
+     * Atomically reads the value reference from the entry (given by entry index "ei")
+     * synchronisation with reading the value version is not ensured in this method
+     * */
+    private long getValueReference(int ei) {
+        return getEntryArrayFieldLong(entryIdx2intIdx(ei), OFFSET.VALUE_REFERENCE);
+    }
+
+    /**
+     * Atomically reads both the value reference and its value (comparing the version).
+     * It does that by using the atomic snapshot technique (reading the version, then the reference,
+     * and finally the version again, checking that it matches the version read previously).
+     * Since a snapshot is used, the LP is when reading the value reference, if the versions match,
+     * otherwise, the operation restarts.
+     *
+     * @param ei The entry of which the reference and version are read
+     * @param version    an output parameter to return the version
+     * @return the read value reference
+     */
+    private long getValueReferenceAndVersion(int ei, int[] version) {
```

Comment:
is there a good reason to use int[] to return the version? can you return a tuple? Pair?

Reply:
int[] is primitive and  just smaller than other objects, using tuple/Pair or Map.Entry wouldn't save the creation of an additional object...